sql-statements: update ddl assertions behavior (#3787)
pingcap/docs: sql-statements/sql-statement-alter-table.md

2 rows in set (0.00 sec)
```

> TiDB supports the ability to assert that DDL changes will use a particular alter algorithm. This is only an assertion, and does not change the actual algorithm which will be used to modify the table. It can be useful if you only want to permit instant DDL changes during the peak hours of your cluster:

Reviewer: ```suggestion
TiDB supports the ability to assert that DDL changes will use a particular `ALTER` algorithm. This is only an assertion, and does not change the actual algorithm which will be used to modify the table. It can be useful if you only want to permit instant DDL changes during the peak hours of your cluster:
```

{{< copyable "sql" >}}

```sql
ALTER TABLE t1 DROP INDEX c1, ALGORITHM=INSTANT;
```

```sql
Query OK, 0 rows affected (0.24 sec)
```

> Using the `INSTANT` assertion on an operation that requires the `INPLACE` algorithm will result in a statement error:

Reviewer: To keep the description consistent with the following one.
```suggestion
Using the `ALGORITHM=INSTANT` assertion on an operation that requires the `INPLACE` algorithm results in a statement error:
```

{{< copyable "sql" >}}

```sql
ALTER TABLE t1 ADD INDEX (c1), ALGORITHM=INSTANT;
```

```sql
ERROR 1846 (0A000): ALGORITHM=INSTANT is not supported. Reason: Cannot alter table by INSTANT. Try ALGORITHM=INPLACE.
```

> However, using the `ALGORITHM=COPY` assertion for an `INPLACE` operation generates a warning instead of an error. This is because TiDB interprets the assertion as _this algorithm or better_. This behavior difference is useful for MySQL compatibility because the algorithm TiDB uses may differ from MySQL:

Reviewer: ```suggestion
However, using the `ALGORITHM=COPY` assertion for an `INPLACE` operation generates a warning instead of an error. This is because TiDB interprets the assertion as _this algorithm or better_. This behavior difference is useful for MySQL compatibility because the algorithm TiDB uses might differ from MySQL:
```

{{< copyable "sql" >}}

```sql
ALTER TABLE t1 ADD INDEX (c1), ALGORITHM=COPY;
SHOW WARNINGS;
```

```sql
Query OK, 0 rows affected, 1 warning (0.25 sec)

+-------+------+---------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                     |
+-------+------+---------------------------------------------------------------------------------------------+
| Error | 1846 | ALGORITHM=COPY is not supported. Reason: Cannot alter table by COPY. Try ALGORITHM=INPLACE. |
+-------+------+---------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

## MySQL compatibility

> The following major restrictions apply to `ALTER TABLE`:

Reviewer: ```suggestion
The following major restrictions apply to `ALTER TABLE` in TiDB:
```

* Multiple operations cannot be completed in a single `ALTER TABLE` statement.

* Lossy changes such as changing from `BIGINT` to `INT` are currently not supported.

* Spatial data types are not supported.

> Further restrictions are documented under [MySQL Compatibility](/mysql-compatibility.md#ddl).

Reviewer: ```suggestion
For further restrictions, see [MySQL Compatibility](/mysql-compatibility.md#ddl).
```
